### PR TITLE
Adding Cosmos 3 to Diffusers

### PR DESCRIPTION
**Pull request by atharvajoshi10**
###  What does this PR do?

  Adds the Cosmos 3 omni pipeline to Diffusers — NVIDIA's unified world foundation model for Physical AI. Cosmos 3 is a
  single Mixture-of-Transformers (MoT) model that combines world generation, physical reasoning, and action generation
  into one forward pass, replacing the separate Predict / Reason / Transfer models from earlier Cosmos releases.

###  The integration ships:

  - `Cosmos3OmniPipeline` — one pipeline class supporting four workflows:
    - text-to-image (num_frames=1)
    - text-to-video
    - image-to-video
    - text+image-to-video-with-sound (when the checkpoint carries a sound tokenizer; enable_sound=True)
  - `Cosmos3OmniTransformer` — MoT backbone running a Qwen-style causal "understanding" stream in parallel with a
  bi-directional "generation" stream over video + (optional) sound latents, joined by a 3D multimodal RoPE.
  - `Cosmos3AVAEAudioTokenizer` — decoder-only audio tokenizer (Oobleck-style Snake1d + weight-norm conv stack, inlined for
  self-containment).
  - Reuses `AutoencoderKLWan` for the video VAE and `UniPCMultistepScheduler` for diffusion — no new scheduler required.
  - `CosmosSafetyChecker` (from cosmos_guardrail) wired up by default per the NVIDIA Open Model License.
  Disable-at-construction (enable_safety_checker=False) and disable-per-call (`enable_safety_check=False`) flags exist for
  test/dev workflows.
  - Two checkpoints on the Hub: nvidia/Cosmos3-Nano (smaller, faster) and nvidia/Cosmos3-Super (larger, higher quality).

  Refactor opportunistically included: promoted LTX2's PyAV-based `encode_video` helper to `diffusers.utils.export_utils` so
  Cosmos 3 (and any future pipeline that needs to mux audio into MP4) can reuse it.
 `diffusers.pipelines.ltx2.export_utils.encode_video` remains as a deprecation shim with a 0.40.0 removal target, so
  existing user code keeps working.

  Docs page `(docs/source/en/api/pipelines/cosmos3.md)` covers all four workflows with per-model <hfoptions> tabs (Nano / Super), plus metadata-template and safety-checker controls. A minimal smoke-test runner lives at `examples/cosmos3/inference_cosmos3.py` with a `--model {nano,super}` flag.

### CODE REVIEW

File: src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py [hunk 1]
```diff
@@ -0,0 +1,972 @@
...
+                        vision_condition_mask=[vision_condition_mask],
+                        sound_condition_mask=[sound_condition_mask] if sound_condition_mask is not None else None,
+                    )
+
+                # --- CFG combine + per-modality scheduler step ---
+                # UniPC's multistep_uni_p_bh_update einsum ("k,bkc...->bc...") requires sample
+                # to carry a batch dim; per-modality latents have no batch axis, so wrap for the step.
+
+                # Skip CFG for 1.0 guidance scale
+                if guidance_scale != 1.0:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py:992]:
```suggestion
                if self.do_classifier_free_guidance:
```

File: src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py [hunk 2]
```diff
@@ -0,0 +1,972 @@
...
+        return torch.from_numpy(checked.astype(np.float32) / 255.0).permute(0, 3, 1, 2)
+
+    @property
+    def current_timestep(self):
+        return self._current_timestep
+
+    @property
+    def interrupt(self):
+        return self._interrupt
+
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py:672]:
```suggestion
  @property
  def guidance_scale(self):
      return self._guidance_scale
      
  @property
  def do_classifier_free_guidance(self):
      return self._guidance_scale > 1.0

```

File: src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py [hunk 3]
```diff
@@ -0,0 +1,972 @@
...
+                else:
+                    velocity_vision = cond_v_vision
+
+                latents = self.scheduler.step(
+                    velocity_vision.unsqueeze(0), t, latents.unsqueeze(0), return_dict=False
+                )[0].squeeze(0)
+
+                if sound_scheduler is not None and cond_v_sound is not None:
+                    # Skip CFG for 1.0 guidance scale
+                    if guidance_scale != 1.0:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py:1003]:
```suggestion
                    if self.do_classifier_free_guidanc:
```

**Review by yiyixuxu — APPROVED** [review]
thanks, I somehow missed the `guidance_scale`
our pipeline has a `do_classifier_freee_guidance` property to indicate if we want to apply CFG or not 
https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/cosmos/pipeline_cosmos2_5_predict.py#L529 
let me know if there is a reason we didn't do that for cosmos3

**Comment by yiyixuxu** [comment]
can you run `make fix-copies`?

**Comment by yiyixuxu** [comment]
@bot /style

File: src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py (hunk 1 above)
Comment (atharvajoshi10) [reply src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py:992]: Addressed in https://github.com/huggingface/diffusers/pull/13823

File: src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py (hunk 2 above)
Comment (atharvajoshi10) [reply src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py:672]: Addressed in https://github.com/huggingface/diffusers/pull/13823

File: src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py (hunk 3 above)
Comment (atharvajoshi10) [reply src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py:1003]: Addressed in https://github.com/huggingface/diffusers/pull/13823

**Comment by MeiYi-dev** [comment]
> ### What does this PR do?
> Adds the Cosmos 3 omni pipeline to Diffusers — NVIDIA's unified world foundation model for Physical AI. Cosmos 3 is a single Mixture-of-Transformers (MoT) model that combines world generation, physical reasoning, and action generation into one forward pass, replacing the separate Predict / Reason / Transfer models from earlier Cosmos releases.
> 
> ### The integration ships:
> * `Cosmos3OmniPipeline` — one pipeline class supporting four workflows:
>   
>   * text-to-image (num_frames=1)
>   * text-to-video
>   * image-to-video
>   * text+image-to-video-with-sound (when the checkpoint carries a sound tokenizer; enable_sound=True)
> * `Cosmos3OmniTransformer` — MoT backbone running a Qwen-style causal "understanding" stream in parallel with a
>   bi-directional "generation" stream over video + (optional) sound latents, joined by a 3D multimodal RoPE.
> * `Cosmos3AVAEAudioTokenizer` — decoder-only audio tokenizer (Oobleck-style Snake1d + weight-norm conv stack, inlined for
>   self-containment).
> * Reuses `AutoencoderKLWan` for the video VAE and `UniPCMultistepScheduler` for diffusion — no new scheduler required.
> * `CosmosSafetyChecker` (from cosmos_guardrail) wired up by default per the NVIDIA Open Model License.
>   Disable-at-construction (enable_safety_checker=False) and disable-per-call (`enable_safety_check=False`) flags exist for
>   test/dev workflows.
> * Two checkpoints on the Hub: nvidia/Cosmos3-Nano (smaller, faster) and nvidia/Cosmos3-Super (larger, higher quality).
> 
> Refactor opportunistically included: promoted LTX2's PyAV-based `encode_video` helper to `diffusers.utils.export_utils` so Cosmos 3 (and any future pipeline that needs to mux audio into MP4) can reuse it. `diffusers.pipelines.ltx2.export_utils.encode_video` remains as a deprecation shim with a 0.40.0 removal target, so existing user code keeps working.
> 
> Docs page `(docs/source/en/api/pipelines/cosmos3.md)` covers all four workflows with per-model tabs (Nano / Super), plus metadata-template and safety-checker controls. A minimal smoke-test runner lives at `examples/cosmos3/inference_cosmos3.py` with a `--model {nano,super}` flag.

Can we also get this model into ComfyUI please so we can use it in our workflows?
